testsupport/wait/host.go

```diff
@@ -162,138 +168,158 @@ func (a *HostAwaitility) UpdateUserSignupSpec(userSignupName string, modifyUserS
+				if !containsUserAccountStatus(actual.Status.UserAccounts, expUaStatus) {
+					return false
+				}
+			}
+			return true
+		},
+		Diff: func(actual *toolchainv1alpha1.MasterUserRecord) string {
+			return fmt.Sprintf("expected UserAccount statuses to match: %s", Diff(actual.Status.UserAccounts, expected))
+		},
 	}
 }
 
 // UserSignupWaitCriterion a function to check that a user account has the expected condition
-type UserSignupWaitCriterion func(a *HostAwaitility, ua *toolchainv1alpha1.UserSignup) bool
+type UserSignupWaitCriterion struct {
+	Match func(a *HostAwaitility, ua *toolchainv1alpha1.UserSignup) bool
+	Diff  func(ua *toolchainv1alpha1.UserSignup) string
+}
 
 // UntilUserSignupHasConditions returns a `UserAccountWaitCriterion` which checks that the given
 // UserAccount has exactly all the given status conditions
-func UntilUserSignupHasConditions(conditions ...toolchainv1alpha1.Condition) UserSignupWaitCriterion {
-	return func(a *HostAwaitility, ua *toolchainv1alpha1.UserSignup) bool {
-		if test.ConditionsMatch(ua.Status.Conditions, conditions...) {
-			a.T.Logf("status conditions match in UserSignup '%s`", ua.Name)
-			return true
-		}
-		a.T.Logf("waiting for status condition of UserSignup '%s'. Actual: '%+v'; Expected: '%+v'", ua.Name, ua.Status.Conditions, conditions)
-		return false
+func UntilUserSignupHasConditions(expected ...toolchainv1alpha1.Condition) UserSignupWaitCriterion {
+	return UserSignupWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.UserSignup) bool {
+			return test.ConditionsMatch(actual.Status.Conditions, expected...)
+		},
+		Diff: func(actual *toolchainv1alpha1.UserSignup) string {
+			return fmt.Sprintf("expected conditions to match:\n%s", Diff(actual.Status.Conditions, expected))
+		},
 	}
 }
 
 // ContainsCondition returns a `UserAccountWaitCriterion` which checks that the given
 // UserAccount contains the given status condition
-func ContainsCondition(condition toolchainv1alpha1.Condition) UserSignupWaitCriterion {
-	return func(a *HostAwaitility, ua *toolchainv1alpha1.UserSignup) bool {
-		if test.ContainsCondition(ua.Status.Conditions, condition) {
-			a.T.Logf("UserSignup '%s' status conditions contains expected condition", ua.Name)
-			return true
-		}
-		a.T.Logf("waiting for status condition of UserSignup '%s' to contain the expected. Actual: '%+v'; Expected: '%+v'", ua.Name, ua.Status.Conditions, condition)
-		return false
+func ContainsCondition(expected toolchainv1alpha1.Condition) UserSignupWaitCriterion {
+	return UserSignupWaitCriterion{
+		Match: func(a *HostAwaitility, actual *toolchainv1alpha1.UserSignup) bool {
+			return test.ContainsCondition(actual.Status.Conditions, expected)
+		},
+		Diff: func(actual *toolchainv1alpha1.UserSignup) string {
+			return fmt.Sprintf("expected conditions to contain: %s.\n\tactual: %s", spew.Sdump(expected), spew.Sdump(actual.Status.Conditions))
```

Comment:
The same question - why don't we use the diff?

Reply:
This line makes me also thinking - wouldn't it be easier to marshal the objects (interfaces) into YAML format and print the expected and actual one? you could also use some tool for comparing the YAMLs.

Reply:
`spew.Sdump` already marshalls the objects. I think that in the case of finding a conditions in a set of conditions, dumping the objects is enough. I would suggest we improve the output of `Diff` in the future if it's not clear enough as it is proposed here.

Reply:
Wouldn't the diff help when it's not immediately obvious? eg. If only the reason is different

Reply:
> Wouldn't the diff help when it's not immediately obvious? eg. If only the reason is different

that's the purpose of the `Diff` func, indeed. 
I suggest with mark this conversation as "resolved" for now

Reply:
ah, I see what you meant now! 🤦‍♂️ 
Let me add a diff to compare the structs and spot the differences more easily ;)

Reply:
done in 9085b05